### PR TITLE
Port add-on to be a hybid WebExtension

### PR DESCRIPTION
**Pull request by mkaply**
Moving to a WebExtension hoping for better existence with multiprocess.

Also, removing the distribution telemetry exclusion.

### CODE REVIEW

File: add-on/webextension/background.js [hunk 1]
```diff
@@ -132,190 +139,105 @@ function getSearchDomainCodes(host) {
...
+ * @param {String} type The type of search (sap/follow-on).
+ * @param {String} extra Any additional parameters (Optional)
  */
-function log(message) {
-  // console.log(message);
+function sendSaveTelemetryMsg(code, sap, type, extra) {
+  browser.runtime.sendMessage(
+  {name: kSaveTelemetryMsg,
+   data: {
+   code: code,
```
Comment (Standard8) [inline add-on/webextension/background.js]: nit: formatting. We generally prefer to do objects like this:

```js
browser.runtime.sendMessage({
  name: kSaveTelemetryMsg,
  data: {
    code,
...
  }
});
```

File: add-on/webextension/manifest.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
+
+
+{
+  "manifest_version": 2,
+  "name": "Test",
+  "version": "0.6",
+  "content_scripts_foo": [
+    {
```
Comment (Standard8) [inline add-on/webextension/manifest.json]: nit: the formatting of this file needs fixing, similar to what I described above.

**Comment by mkaply** [comment]
The reason I'm switching to a WebExtension is because I want to remove framescripts and the unknowns they introduce when combined with multiprocess.

The data seems to indicate that when Firefox 55 was switched to 4 content process, bad things started happening in the add-on. While I haven't been able to recreate, I'm just not convinced that framescripts weren't at fault here.

The WebExtension model is a much cleaner model for what we are doing here and I think it will give us better results long term.

**Comment by mkaply** [comment]
To give a more concrete answer, listening to loadProgress is giving too many false positives (like new tab thumbnails for instance, and lazy loading tabs).

In order for this to work correctly, we need to look at real page loads in the browser, and the best way to do that is via WebExtensions. Everything they have focuses on real page loads.

In addition, with fixes to WebExtensions in Firefox, we'll be able to tell when a page load is because of a lazy  loading tab which will avoid false positives.

File: add-on/bootstrap.js [hunk 3]
```diff
@@ -69,70 +65,6 @@ function handleSaveTelemetryMsg(message) {
...
-  init() {
-    if (this._definedThisSession) {
-      return;
-    }
-
-    this._definedThisSession = true;
-    this.enableForUser = false;
-
-    try {
-      let distId = Services.prefs.getCharPref("distribution.id", "");
```
Comment (Standard8) [inline add-on/bootstrap.js]: Did you mean to stop checking to see if this was a distribution or not? If you just want to simplify the code that'd be fine.

File: add-on/bootstrap.js [hunk 4]
```diff
@@ -152,6 +84,8 @@ function uninstall(data, reason) {
   // Nothing specifically to do, shutdown does what we need.
 }
 
+let addonResourceURI;
```
Comment (Standard8) [inline add-on/bootstrap.js]: This doesn't seem to be used.

File: add-on/webextension/contentscript.js [hunk 5]
```diff
@@ -0,0 +1,174 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (Standard8) [inline add-on/webextension/contentscript.js]: This file seems like an accidentally created duplicate? It doesn't seem to be referenced.

File: add-on/bootstrap.js (hunk 3 above)
Comment (mkaply) [reply add-on/bootstrap.js]: Yes. In the end we realized that we want all the data because that's what's being reported in telemetry.

**Comment by mkaply** [comment]
I think I've addressed your changes (and merged the latest code.).

**Comment by Standard8** [comment]
For reference, I've just pushed to try server to get a Talos performance comparison, here's the link:

https://treeherder.mozilla.org/perf.html#/compare?originalProject=try&originalRevision=364051528b69a0ebbbc4c27f545b4256487bf36f&newProject=try&newRevision=235334fcb04f470e124c82de904903526b9afdeb&framework=1&filter=tabpa

It'll probably be a few hours before we get all results in.

**Comment by Standard8** [comment]
The comparison looks good, possibly even a slight perf win on win64 tpaint e10s. I realised to be safe we should look at PGO builds as well:

https://treeherder.mozilla.org/perf.html#/compare?originalProject=try&originalRevision=364051528b69a0ebbbc4c27f545b4256487bf36f&newProject=try&newRevision=bc23a93347fc8e9e1dbaa52588859f91fd903e3e&framework=1&showOnlyImportant=1

**Comment by Standard8** [comment]
@mkaply The latest comparison looks good, so I think we are a go from a perf perspective.

I pushed a few extra changes for fixing up lint and the unit tests. Can you take a quick look over them - especially the add-on affecting parts - and just check they look ok?

I'm happy to try and rebase or somehow get this merged to master once you've had a look.

File: add-on/bootstrap.js [hunk 6]
```diff
@@ -103,6 +103,9 @@ function startup(data, reason) {
   webExtension.startup(reason).then(api => {
     const {browser} = api;
     browser.runtime.onMessage.addListener(handleSaveTelemetryMsg);
+    return Promise.resolve();
```
Comment (mkaply) [inline add-on/bootstrap.js]: I don't know enough about promises to understand what this is. Can you explain? Tx.

**Comment by mkaply** [comment]
This looks good to me.

File: add-on/bootstrap.js (hunk 6 above)
Comment (Standard8) [reply add-on/bootstrap.js]: This returns a [promise that is already resolved](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve). In this case, we're ensuring good practice by returning a promise from a promise handler.

**Comment by Standard8** [comment]
Thanks, I'll try and merge this.

**Pull request merged**